File: lib/features/subscriptions/models/subscription.dart
import 'package:subtracker/core/extensions/datetime_extensions.dart';
import 'package:subtracker/features/subscriptions/models/billing_cycle.dart';
import 'package:subtracker/features/subscriptions/models/subscription_category.dart';

class Subscription {
  const Subscription({
    required this.id,
    required this.name,
    this.description,
    required this.amount,
    this.currency = 'EUR',
    required this.billingCycle,
    required this.category,
    required this.startDate,
    required this.nextBillingDate,
    this.isActive = true,
    this.url,
    this.reminderDaysBefore = 2,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final String? description;
  final double amount;
  final String currency;
  final BillingCycle billingCycle;
  final SubscriptionCategory category;
  final DateTime startDate;
  final DateTime nextBillingDate;
  final bool isActive;
  final String? url;
  final int reminderDaysBefore;
  final DateTime createdAt;
  final DateTime updatedAt;

  double get monthlyAmount => billingCycle.monthlyEquivalent(amount);
  double get yearlyAmount => billingCycle.yearlyEquivalent(amount);
  int get daysUntilNextBilling => nextBillingDate.daysFromNow;
  bool get isDueSoon => isActive && daysUntilNextBilling <= reminderDaysBefore;

  factory Subscription.fromJson(Map<String, dynamic> json) {
    return Subscription(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] as String,
      billingCycle: BillingCycle.fromJson(json['billingCycle'] as int),
      category: SubscriptionCategory.fromJson(json['category'] as int),
      startDate: DateTime.parse(json['startDate'] as String),
      nextBillingDate: DateTime.parse(json['nextBillingDate'] as String),
      isActive: json['isActive'] as bool,
      url: json['url'] as String?,
      reminderDaysBefore: json['reminderDaysBefore'] as int,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'amount': amount,
      'currency': currency,
      'billingCycle': billingCycle.toJson(),
      'category': category.toJson(),
      'startDate': startDate.toIso8601String(),
      'nextBillingDate': nextBillingDate.toIso8601String(),
      'isActive': isActive,
      'url': url,
      'reminderDaysBefore': reminderDaysBefore,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  Subscription copyWith({
    String? id,
    String? name,
    String? description,
    double? amount,
    String? currency,
    BillingCycle? billingCycle,
    SubscriptionCategory? category,
    DateTime? startDate,
    DateTime? nextBillingDate,
    bool? isActive,
    String? url,
    int? reminderDaysBefore,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Subscription(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      billingCycle: billingCycle ?? this.billingCycle,
      category: category ?? this.category,
      startDate: startDate ?? this.startDate,
      nextBillingDate: nextBillingDate ?? this.nextBillingDate,
      isActive: isActive ?? this.isActive,
      url: url ?? this.url,
      reminderDaysBefore: reminderDaysBefore ?? this.reminderDaysBefore,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: api/src/SubTracker.Api/Features/Subscriptions/Domain/Subscription.cs
namespace SubTracker.Api.Features.Subscriptions.Domain;

public sealed class Subscription
{
    public Guid Id { get; private set; }
    public string Name { get; private set; } = string.Empty;
    public string? Description { get; private set; }
    public decimal Amount { get; private set; }
    public string Currency { get; private set; } = "EUR";
    public BillingCycle BillingCycle { get; private set; }
    public SubscriptionCategory Category { get; private set; }
    public DateTime StartDate { get; private set; }
    public DateTime NextBillingDate { get; private set; }
    public bool IsActive { get; private set; } = true;
    public string? Url { get; private set; }
    public int ReminderDaysBefore { get; private set; } = 2;
    /// <summary>
    /// Last time a notification was sent for this subscription.
    /// Used to prevent duplicate notifications within the same billing cycle.
    /// </summary>
    public DateTime? LastNotifiedAt { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime UpdatedAt { get; private set; }

    private Subscription() { }

    public static Subscription Create(
        string name,
        string? description,
        decimal amount,
        string currency,
        BillingCycle billingCycle,
        SubscriptionCategory category,
        DateTime startDate,
        string? url,
        int reminderDaysBefore,
        DateTime utcNow)
    {
        return new Subscription
        {
            Id = Guid.NewGuid(),
            Name = name,
            Description = description,
            Amount = amount,
            Currency = currency,
            BillingCycle = billingCycle,
            Category = category,
            StartDate = startDate,
            NextBillingDate = CalculateNextBillingDate(startDate, billingCycle, utcNow),
            Url = url,
            ReminderDaysBefore = reminderDaysBefore,
            CreatedAt = utcNow,
            UpdatedAt = utcNow
        };
    }

    public void Update(
        string name,
        string? description,
        decimal amount,
        string currency,
        BillingCycle billingCycle,
        SubscriptionCategory category,
        DateTime startDate,
        string? url,
        int reminderDaysBefore,
        bool isActive,
        DateTime utcNow)
    {
        Name = name;
        Description = description;
        Amount = amount;
        Currency = currency;
        BillingCycle = billingCycle;
        Category = category;
        StartDate = startDate;
        NextBillingDate = CalculateNextBillingDate(startDate, billingCycle, utcNow);
        Url = url;
        ReminderDaysBefore = reminderDaysBefore;
        IsActive = isActive;
        UpdatedAt = utcNow;
    }

    public void Deactivate(DateTime utcNow)
    {
        IsActive = false;
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Returns true if the subscription is active and NextBillingDate is within
    /// the reminder window (today &lt;= NextBillingDate &lt;= today + ReminderDaysBefore).
    /// Excludes subscriptions with NextBillingDate in the past.
    /// </summary>
    public bool IsDueSoon(DateTime utcNow)
    {
        if (!IsActive) return false;

        var daysUntilBilling = (NextBillingDate.Date - utcNow.Date).TotalDays;
        return daysUntilBilling >= 0 && daysUntilBilling <= ReminderDaysBefore;
    }

    /// <summary>
    /// Determines if a notification should be sent for this subscription.
    /// Returns true if the subscription is due soon AND has not been notified
    /// in the current billing cycle.
    /// </summary>
    public bool NeedsNotification(DateTime utcNow)
    {
        if (!IsDueSoon(utcNow)) return false;

        if (LastNotifiedAt is null) return true;

        var cycleStart = GetCycleStartDate();
        return LastNotifiedAt.Value < cycleStart;
    }

    /// <summary>
    /// Marks this subscription as having been notified.
    /// </summary>
    public void MarkNotified(DateTime utcNow)
    {
        LastNotifiedAt = utcNow;
    }

    /// <summary>
    /// Returns true if NextBillingDate is in the past relative to utcNow.
    /// </summary>
    public bool HasPastBillingDate(DateTime utcNow) =>
        IsActive && NextBillingDate.Date < utcNow.Date;

    /// <summary>
    /// Advances NextBillingDate to the next valid future date based on the billing cycle.
    /// </summary>
    public void AdvanceNextBillingDate(DateTime utcNow)
    {
        NextBillingDate = CalculateNextBillingDate(NextBillingDate, BillingCycle, utcNow);
    }

    /// <summary>
    /// Calculates the start of the current billing cycle.
    /// If LastNotifiedAt is before this date, a new notification is needed.
    /// </summary>
    private DateTime GetCycleStartDate()
    {
        return BillingCycle switch
        {
            BillingCycle.Weekly => NextBillingDate.AddDays(-7),
            BillingCycle.Monthly => NextBillingDate.AddMonths(-1),
            BillingCycle.Quarterly => NextBillingDate.AddMonths(-3),
            BillingCycle.Yearly => NextBillingDate.AddYears(-1),
            _ => NextBillingDate.AddMonths(-1)
        };
    }

    private static DateTime CalculateNextBillingDate(
        DateTime startDate,
        BillingCycle cycle,
        DateTime utcNow)
    {
        var next = startDate;
        while (next <= utcNow)
        {
            next = cycle switch
            {
                BillingCycle.Weekly => next.AddDays(7),
                BillingCycle.Monthly => next.AddMonths(1),
                BillingCycle.Quarterly => next.AddMonths(3),
                BillingCycle.Yearly => next.AddYears(1),
                _ => next.AddMonths(1)
            };
        }
        return next;
    }
}
